src/main/java/org/opensearch/neuralsearch/processor/combination/ScoreCombiner.java

```diff
@@ -61,34 +96,177 @@ private void combineShardScores(final ScoreCombinationTechnique scoreCombination
+     * @return list of top field docs which is deduced by typcasting top docs to top field docs.
+     */
+    private List<TopFieldDocs> getTopFieldDocs(final Sort sort, final List<TopDocs> topDocsPerSubQuery) {
+        if (sort == null) {
+            return null;
+        }
+        List<TopFieldDocs> topFieldDocs = new ArrayList<>();
+        for (TopDocs topDocs : topDocsPerSubQuery) {
+            // Check for scoreDocs length.
+            // If scoreDocs length=0 then it means that no results are found for that particular subquery.
+            if (topDocs.scoreDocs.length != 0) {
+                topFieldDocs.add((TopFieldDocs) topDocs);
+            }
+        }
+        return topFieldDocs;
+    }
+
+    /**
+     *
+      * @param compoundTopDocs top docs that represent on shard
+     * @param combinedNormalizedScoresByDocId docId to normalized scores map
+     * @param sort sort criteria
+     * @return map of docId and sort fields if sorting is enabled.
+     */
+    private Map<Integer, Object[]> getDocIdSortFieldsMap(
+        final CompoundTopDocs compoundTopDocs,
+        final Map<Integer, Float> combinedNormalizedScoresByDocId,
+        final Sort sort
+    ) {
+        // If sort is null then no sort fields present therefore return null.
+        if (sort == null) {
+            return null;
+        }
         // we're merging docs with normalized and combined scores. we need to have only maxHits results
-        List<Integer> sortedDocsIds = new ArrayList<>(combinedNormalizedScoresByDocId.keySet());
-        sortedDocsIds.sort((a, b) -> Float.compare(combinedNormalizedScoresByDocId.get(b), combinedNormalizedScoresByDocId.get(a)));
-        return sortedDocsIds;
+        Map<Integer, Object[]> docIdSortFieldMap = new HashMap<>();
+        List<TopDocs> topFieldDocs = compoundTopDocs.getTopDocs();
+        final boolean isSortByScore = checkIfSortOrderByScore(sort);
+        for (TopDocs topDocs : topFieldDocs) {
+            for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+                FieldDoc fieldDoc = (FieldDoc) scoreDoc;
+
+                if (docIdSortFieldMap.get(fieldDoc.doc) == null) {
+                    // If sort by score then replace sort field value with normalized score.
+                    if (isSortByScore) {
+                        docIdSortFieldMap.put(fieldDoc.doc, new Object[] { combinedNormalizedScoresByDocId.get(fieldDoc.doc) });
+                    } else {
+                        docIdSortFieldMap.put(fieldDoc.doc, fieldDoc.fields);
+                    }
+                }
+            }
+        }
+        return docIdSortFieldMap;
+    }
+
+    private Collection<Integer> getSortedDocIds(
+        final Map<Integer, Float> combinedNormalizedScoresByDocId,
+        final List<TopFieldDocs> topFieldDocs,
+        final Sort sort
+    ) {
+        // we're merging docs with normalized and combined scores. we need to have only maxHits results
+        if (Objects.isNull(sort)) {
```

Comment:
nit: some place i see == null check is added and here we are using Objects api. can you make this as consistent? i don't have any preference, just want to see we are using one format to perform null check.